### ISSUE TITLE
builtins support

### ISSUE THREAD

**Issue by mental32** [mental32/monty#7]
~~Currently we don't have any importing mechanisms or module infrastructure but this shouldn't matter _too_ much for the bullt-ins since people don't typically use them through the `builtins` module import.~~

~~We now have minimal functioning support for modules & importing!
At the moment builtins have to be explicitly imported but we can fix this easily enough :)~~

Big list of stuff to implement:

- [ ] abs
- [ ] all
- [ ] any
- [ ] ascii
- [ ] bin
- [x] bool
- [ ] breakpoint (I don't have a god damn clue how to deal with this one)
- [ ] bytearray
- [ ] callable (Actually this should be fairly simple!)
- [ ] chr
- [ ] classmethod
- [ ] compile
- [ ] complex
- [ ] delattr
- [ ] dict
- [ ] dir
- [ ] dimod
- [ ] enumerate
- [ ] filter
- [x] float
- [ ] format
- [ ] frozenset
- [ ] getattr
- [ ] globals (comptime only)
- [ ] hasattr (comptime only?)
- [ ] hash
- [ ] hex
- [x] id (this is basically the "address of")
- [ ] input
- [x] int (a zero argument version of this should be easy asf)
- [x] isinstance(comptime only?)
- [ ] issubclass (comptime only?)
- [ ] iter (_shrug_)
- [ ] len (ehhh kinda UB)
- [ ] list
- [ ] locals (comptime only)
- [ ] map (_shrug_)
- [ ] max
- [ ] memoryview (what was this?)
- [ ] min
- [ ] next (_shrug_)
- [x] object (_big shrug_)
- [ ] oct
- [ ] open
- [ ] ord
- [ ] pow
- [ ] print (big oof, nearly done)
- [ ] property (_SHRUG_)
- [ ] range
- [ ] repr
- [ ] reversed
- [ ] round
- [ ] set
- [ ] setattr
- [ ] slice
- [ ] sorted
- [ ] staticmethod
- [x] str (lmao no, we need heap allocation for this baby)
- [ ] sum
- [ ] super
- [ ] tuple
- [x] type
- [ ] vars
- [ ] zip
- [ ] `__import__`

**Comment by mental32** [comment]
As of #9 we have the necessary import machinery in order to describe and import the "builtins" modules.

Now all that's left is to (high-level overview):

* Implement the builtins modules
* Implement the individual builtins above.

Easier said than done!

Take for example the `print` builtin, at the moment the current implementation will look something like this:

```py
print: Callable[[str], NoneType]
```

The ~~laziest~~ _easiest_ solution at the moment is to only take a single string argument and to write it to stdout. Ideally like to account for the possibility of multiple operating systems although at the moment I can only test support for unix-like systems so this will probably be the only real implementation of `print` for a while.

I like the idea of being able to behave at compile time, similar to how Zig does with its `comptime` blocks. So we could perform DCE on constants exported through `__monty` and therefore expose the correct information available at compile time.

```py
# Implentation of stdlib/builtins/print_impl.py
import __monty

if __monty.arch.os == "unix":
    def print(st: str):
        # sys.write(st, STDOUT)
else:
    assert False, "boom!"
```

This means that we would need to expose this "magic" module (we could probably piggyback this abstraction through the sys module) and then correctly perform the compile time constant time operations and even _then_ we need to perform the appropriate DCE before lowering into MIR, so at the ast level.

Basically its going to take a little while...

**Comment by mental32** [comment]
So I've managed to get some IR that forms accordingly; disassembly:
```
data:
  s0 = Module(name='builtins', path=PosixPath('/home/mental/Desktop/open_projects/monty/stdlib/builtins'))
  s1 = 'Hello, World!'

builtins:
  func print({}) -> none_
    b0:
      nop;

__main__:
  func main({}) -> none_
    r0 = <Function: signature='print() -> None'>
    b0:
      v0 = str.const(s1);
      v1 = call r0(v0);
```

We're actually not that far off from a hello world IMO, this IR should in its current state **not** typecheck correctly since print is defined as `Callable[[], None]` and we're passing in a string constant as an argument that's something that will get sorted out eventually. 

The more interesting question is getting the capability to describe symbols to "import", we namely need `libc` to help us here and we need to reference `FILE *stdout` and `fwrite`. It's perfectly doable just tedious at the moment, but a "real" hello world with maybe a primitive MIRI should come in sometime in the middle of next October :smile:

**Comment by mental32** [comment]
With the latest change-set IR now forms accordingly; disassembly:
```
data:
  s0 = Module(name='builtins', path=PosixPath('/home/mental/Desktop/open_projects/monty/stdlib/builtins'))
  s1 = 'w'
  s2 = 'Hello, World!'

__monty: <empty>
builtins:
  func print({}) -> none_
    r0 = <ExternalFunction: signature='fopen(<Primitive.I32: 12>, <Primitive.Str: 6>) -> <Primitive.Int: 10>'>
    r1 = <ExternalFunction: signature='fwrite(<Primitive.I32: 12>, <Primitive.I32: 12>, <Primitive.I32: 12>, <Primitive.I32: 12>) -> <Primitive.Int: 10>'>
    r2 = <Function: signature='len(st: str) -> int'>
    r3 = <ExternalFunction: signature='fflush(<Primitive.I32: 12>) -> <Primitive.I32: 12>'>
    b0:
      v1 = iconst.64 1;
      v2 = str.const(s1);
      v3 = call r0(v1, v2);
      BlockInstr(op=<InstrOp.StackStore: 9>, args=[4, 3], ret=None);
      BlockInstr(op=<InstrOp.StackLoad: 10>, args=[0, 0, 6], ret=5);
      v6 = iconst.64 1;
      BlockInstr(op=<InstrOp.StackLoad: 10>, args=[0, 0, 6], ret=7);
      v8 = call r2(v7);
      BlockInstr(op=<InstrOp.StackLoad: 10>, args=[4, 4, 10], ret=9);
      v10 = call r1(v5, v6, v8, v9);
      BlockInstr(op=<InstrOp.StackLoad: 10>, args=[4, 4, 10], ret=11);
      v12 = call r3(v11);

  func len({}) -> int
    b0:
      v1 = iconst.64 20;
      return v1;

  

__main__:
  func main({}) -> none_
    r0 = <Function: signature='print(st: str) -> None'>
    b0:
      v0 = str.const(s2);
      v1 = call r0(v0);
```

We're on the right track atm, currently what's missing is some form of compile time string-constant analysis. Notice the implementation of the `len` builtin has been fixed to return the value `20` (this makes it officially UB). Actually any form of compile time behaviour is pretty much impossible at the moment... I suggest starting at those optimizations after we lower the AST into MIR since it'll save us time re-coding and sanity typechecking every root tree of every module we describe.

So the next steps are I assume as follows:

* Implement some official way to describe compile time behavior, like static asserts or folding branches where the condition is known at compile time such as the target architecture being compiled for.
* Enhance the typechecker both at the AST/Item level and the miniature MIR sanity value typechecker
* Get this IR to something capable of executing it (probably MIRI)

**Comment by mental32** [comment]
Triage: `len` is not as UB as before, the function will just invoke `obj.__len__()` which is only currently implemented on `str` to return a fixed `0`

We do now have `comptime` however! it's essentially just a AST pre-processor that kicks in immedietly after parsing but before everything else.
